src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java

```diff
@@ -0,0 +1,195 @@
+            if (hasKafkaConnectionException(exceptions)) {
+                throw eventPublishingException;
+            }
+            return new CommandResult(eventPublishingException);
+        }
+        return new CommandResult();
+    }
+
+    private boolean hasKafkaConnectionException(final List<Exception> exceptions) {
+        return exceptions.stream().filter(ex ->
+                ex instanceof org.apache.kafka.common.errors.TimeoutException ||
+                ex instanceof NetworkException ||
+                ex instanceof UnknownServerException)
+                .findFirst().isPresent();
+    }
+
+    private static boolean isExceptionShouldLeadToReset(@Nullable final Exception exception) {
+        if (null == exception) {
+            return false;
+        }
+        return Stream.of(NotLeaderForPartitionException.class, UnknownTopicOrPartitionException.class).
+                filter(clazz -> clazz.isAssignableFrom(exception.getClass()))
+                .findAny().isPresent();
+    }
+
+    private void failUnpublished(final List<BatchItem> batch, final String reason) {
+        batch.stream()
+                .filter(item -> item.getResponse().getPublishingStatus() != EventPublishingStatus.SUBMITTED)
+                .forEach(item -> item.updateStatusAndDetail(EventPublishingStatus.FAILED, reason));
+    }
+
+    private static CompletableFuture<Exception> publishItem(
+            final Producer<String, String> producer, final String topicId, final BatchItem item)
+            throws EventPublishingException {
+        try {
+            final CompletableFuture<Exception> result = new CompletableFuture<>();
+            final ProducerRecord<String, String> kafkaRecord = new ProducerRecord<>(
+                    topicId,
+                    KafkaCursor.toKafkaPartition(item.getPartition()),
+                    item.getPartition(),
+                    item.getEvent().toString());
+
+            producer.send(kafkaRecord, ((metadata, exception) -> {
+                if (null != exception) {
+                    item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+                    result.complete(exception);
+                } else {
+                    item.updateStatusAndDetail(EventPublishingStatus.SUBMITTED, "");
+                    result.complete(null);
+                }
+            }));
+            return result;
+        } catch (final InterruptException e) {
+            item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+            throw new EventPublishingException("Error publishing message to kafka", e);
+        } catch (final RuntimeException e) {
+            item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+            throw new EventPublishingException("Error publishing message to kafka", e);
+        }
+    }
+
+    public static class CommandResult {
+
+        private final EventPublishingException exception;
```

Comment:
@rcillo I thought about it but there are no so many cases for that. It is very specific case when we have to return one exception and throw another. I think it is easier to understand with less code.